File: lib/Jordan-Gauss/Jordan_Gauss.fs
//Программа для расчета линейных уравнений по методу Жордана-Гаусса
//На вход приходит матрица коэффициентов и массив значений
//На выходе возвращается сама система и её результат
module Jordan_Gauss

    type public Jordan_Gauss = {Table:System.Double[,] ; ResultTable : System.Double[,] }

    let public StepMethod (data:Jordan_Gauss, indexRow:int, indexColumn:int) : Jordan_Gauss =
        let mutable res = {Table = data.Table;ResultTable = data.ResultTable  }

        if res.ResultTable = null then
            res<-{Table = data.Table;ResultTable = Array2D.create (data.Table.GetLength(0)) (data.Table.GetLength(1)) 0.0}
        
        if res.Table.[indexRow,indexColumn] <> 1.0 then
           let firstElem = res.Table.[indexRow,indexColumn]
           for indexColumn=0 to res.Table.GetLength(1) - 1 do
                if firstElem <> 0.0 
                then  Array2D.set res.ResultTable indexRow indexColumn (res.Table.[indexRow,indexColumn] / firstElem)
                else Array2D.set res.ResultTable indexRow indexColumn 0.0

        for row=0 to res.Table.GetLength(0) - 1 do
            if row <> indexRow then
                let firstElem = res.Table.[row,indexColumn]
                for column=indexColumn to res.Table.GetLength(1) - 1 do
                    Array2D.set res.ResultTable row column (res.Table.[row,column] - (firstElem * res.ResultTable.[indexRow,column]))
        res <- {Table = res.ResultTable;ResultTable = res.ResultTable  }
        
        res
            
            
    let public RunMethod (data:Jordan_Gauss) : Jordan_Gauss =
        let mutable res = {Table = data.Table;ResultTable = Array2D.create (data.Table.GetLength(0)) (data.Table.GetLength(1)) 0.0}
        for row = 0 to data.Table.GetLength(0) - 1 do
            printfn "%i" row
            res <- StepMethod(res,row,row)   
        res


    let public PrintJordan_Gauss (data:Jordan_Gauss) : string =
        data.ToString()

File: test/JordanGaussTest/JordanGauss.cs
using System;
using NUnit.Framework;

namespace JordanGaussTest
{
    [TestFixture]
    public class JordanGauss
    {
        [Test]
        public void Test_FullJordanGauss()
        {
            //шаг 1
            double[,] except =
                {{0, 1, -2, 3}, {1, -2, 5.0 / 2, -13.0 / 2}, {0, 5, 17.0 / 2, -7.0 / 2}};
            Jordan_Gauss.Jordan_Gauss actual = Jordan_Gauss.StepMethod
            (
                data: new Jordan_Gauss.Jordan_Gauss
                (
                    table: new double[,] {{4, -7, 8, -23}, {2, -4, 5, -13}, {-3, 11, 1, 16}},
                    resultTable: null
                ),
                indexRow: 1,
                indexColumn: 0
            );
            Console.WriteLine(value: Jordan_Gauss.PrintJordan_Gauss(data: actual));
            CollectionAssert.AreEqual(expected: except, actual: actual.ResultTable);

            //шаг 2
            except = new[,] {{0, 1, -2, 3}, {1, 0, -1.5D, -0.5D}, {0, 0, 18.5D, -18.5D}};
            actual = Jordan_Gauss.StepMethod
            (
                data: new Jordan_Gauss.Jordan_Gauss
                    (table: actual.Table, resultTable: actual.ResultTable),
                indexRow: 0,
                indexColumn: 1
            );
            Console.WriteLine(value: Jordan_Gauss.PrintJordan_Gauss(data: actual));
            CollectionAssert.AreEqual(expected: except, actual: actual.ResultTable);

            //шаг 3
            except = new double[,] {{0, 1, 0, 1}, {1, 0, 0, -2}, {0, 0, 1, -1}};
            actual = Jordan_Gauss.StepMethod
            (
                data: new Jordan_Gauss.Jordan_Gauss
                    (table: actual.Table, resultTable: actual.ResultTable),
                indexRow: 2,
                indexColumn: 2
            );
            Console.WriteLine(value: Jordan_Gauss.PrintJordan_Gauss(data: actual));
            CollectionAssert.AreEqual(expected: except, actual: actual.ResultTable);

            //Метод с подным расчетом
            except = new double[,] {{1, 0, 0, -2}, {0, 1, 0, 1}, {0, 0, 1, -1}};
            actual = Jordan_Gauss.RunMethod
            (
                data: new Jordan_Gauss.Jordan_Gauss
                (
                    table: new double[,] {{4, -7, 8, -23}, {2, -4, 5, -13}, {-3, 11, 1, 16}},
                    resultTable: null
                )
            );
            Console.WriteLine(value: Jordan_Gauss.PrintJordan_Gauss(data: actual));
            CollectionAssert.AreEqual(expected: except, actual: actual.ResultTable);
        }

        [Test]
        public void Test_StepJordanGauss()
        {
            //Проверка запуска воозе шага метода
            double[,] except = new double[3, 4];
            Jordan_Gauss.Jordan_Gauss actual = Jordan_Gauss.StepMethod
            (
                data: new Jordan_Gauss.Jordan_Gauss(table: new double[3, 4], resultTable: null),
                indexRow: 1,
                indexColumn: 0
            );
            CollectionAssert.AreEqual(expected: except, actual: actual.ResultTable);
            Console.WriteLine(value: Jordan_Gauss.PrintJordan_Gauss(data: actual));

            //Шаг метода
            except = new[,]
                {{0, 1, -2, 3}, {1, -2, 5.0 / 2, -13.0 / 2}, {0, 5, 17.0 / 2, -7.0 / 2}};
            actual = Jordan_Gauss.StepMethod
            (
                data: new Jordan_Gauss.Jordan_Gauss
                (
                    table: new double[,] {{4, -7, 8, -23}, {2, -4, 5, -13}, {-3, 11, 1, 16}},
                    resultTable: null
                ),
                indexRow: 1,
                indexColumn: 0
            );
            Console.WriteLine(value: Jordan_Gauss.PrintJordan_Gauss(data: actual));
            CollectionAssert.AreEqual(expected: except, actual: actual.ResultTable);
        }
    }
}